### PR TITLE
[Enhancement] Estimate scan row bytes dynamically in time

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [x] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Problem Summary：

The current mechanism of estimating the scanning column bytes is to directly give a fixed estimated value. If the estimated value deviates greatly from the actual value, the parallelism will be too low.
For example, with a `Bitmap` estimate of 1MB, `query_mem_limit` of 32GB would limit the parallelism to 1.

Therefore, this PR adds the strategy of "adaptive row bytes estimation". 
During the scan tasks running, the `estimated_row_bytes` is adjusted in real time according to the read chunk bytes, and then the parallelism is adjusted.

In addition, there is a bug when `max_concurrency` < `dop`, which may lose reading some rows.
```C++
    for (int i = 0; i < MAX_IO_TASKS_PER_OP; ++i) {
        if (_chunk_sources[i] != nullptr && !_is_io_task_running[i] && _chunk_sources[i]->has_next_chunk()) { // A. 
            RETURN_IF_ERROR(_trigger_next_scan(state, i)); // B.
        }
    }
    // Secondly, find the unused position of _chunk_sources to pick up a new morsel.
    if (!_morsel_queue->empty()) {
        for (int i = 0; i < MAX_IO_TASKS_PER_OP; ++i) {
            // *** After limiting the scan concurrency from 2.3, `_trigger_next_scan` at B may failed. 
            // Therefore, should check `!_chunk_sources[i]->has_next_buffer()` again here. ***
            if (_chunk_sources[i] == nullptr || (!_is_io_task_running[i] && !_chunk_sources[i]->has_output())) {
                RETURN_IF_ERROR(_pickup_morsel(state, i));
            }
        }
    }
```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by ZiheLiu** [comment]
@Mergifyio backport branch-2.3

File: be/src/exec/pipeline/scan/scan_operator.h [hunk 1]
```diff
@@ -61,6 +61,11 @@ class ScanOperator : public SourceOperator {
         return res;
     }
 
+    virtual size_t max_scan_concurrency() const {
+        // It takes effect, only when it is positive.
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/scan_operator.h:65]: For `ConnectorScanOperator`, is there any scan concurrency limitation?

**Review by murphyatwork — APPROVED** [review]

File: be/src/exec/pipeline/scan/scan_operator.h (hunk 1 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/scan_operator.h:65]: No, it doesn't have any limitation for now.

**Review by satanson — APPROVED** [review]

**Pull request merged**